### DISCUSSION TITLE
Issues with IPv4 tunneled in IPv6

### DISCUSSION THREAD

**Discussion by hgl** [strongswan/strongswan#850] [General]
I have a working strongswan config where my clients can connect to the IKEv2 server's public IPv4, get a virtual IPv4 address, and access to the network behind the server:

```
connections {
	main {
		version = 2
		pools = main
		proposals = aes128gcm16-prfsha256-ecp256,aes128gcm16-prfsha256-x25519
		send_certreq = no
		childless = never
		if_id_in = 357
		if_id_out = 357
		local {
			auth = pubkey
		}
		remote {
			auth = pubkey
		}
		children {
			child {
				local_ts = 0.0.0.0/0
				esp_proposals = aes128gcm16-ecp256,aes128gcm16-x25519
			}
		}
	}
}

pools {
	main {
		addrs = 192.168.3.100-192.168.3.249
		dns = 192.168.3.1
	}
}
```

However, if a client connects to the IKEv2 server's public IPv6, and get a virtual IPv4 address, it can no longer access to the network behind the server. Tcpdump on the server says the clients gets ICMPv6 destination unreachable error.

The packets inside ICMPv6's ESP are the DNS queries the client sends using the virtual IPv4 as source address and 192.168.3.1 as the destination.

`iptables -t raw -I PREROUTING -d 192.168.3.1 -j TRACE` never captures any packets.

I wonder if tunneling IPv4 inside IPv6 is supported for XFRM interfaces?

**Comment by tobiasbrunner** [comment]
Who sends the ICMPv6? And as response to what message? ESP? The packet inside? (The latter would seem strange because why would the response be IPv6 if the request is IPv4.)

**Reply by hgl** [reply]
Thanks for the quick reply.

The ICMPv6 packets are returned outside IPSec, from the server's public IPv6 to the client's IPv6.

A ICMPv6 response contains the IP packet from the client's IPv6 to the server's public IPv6, which in turn contains the ESP, and which finally contains the DNS query from the client's virtual IPv4 to 192.168.3.1.

Does it sound like a firewall issue? The client can establish the IKEv2 VPN to the server without issues though.

**Reply by tobiasbrunner** [reply]
Yeah, sounds like the server does not accept ESP for IPv6.

**Reply by hgl** [reply]
Thanks, that's indeed the root cause.
